x/warp/contract.go

```diff
@@ -0,0 +1,278 @@
+	// hit an error during execution.
+	unpackedPredicateBytes, err := predicateutils.UnpackPredicate(predicateBytes)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidPredicateBytes, err)
+	}
+	warpMessage, err := warp.ParseMessage(unpackedPredicateBytes)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidWarpMsg, err)
+	}
+
+	addressedPayload, err := warpPayload.ParseAddressedPayload(warpMessage.UnsignedMessage.Payload)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidAddressedPayload, err)
+	}
+	packedOutput, err := PackGetVerifiedWarpMessageOutput(GetVerifiedWarpMessageOutput{
+		Message: WarpMessage{
+			OriginChainID:       warpMessage.SourceChainID,
+			OriginSenderAddress: addressedPayload.SourceAddress,
+			DestinationChainID:  warpMessage.DestinationChainID,
+			DestinationAddress:  addressedPayload.DestinationAddress,
+			Payload:             addressedPayload.Payload,
+		},
+		Exists: true,
+	})
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// UnpackSendWarpMessageInput attempts to unpack [input] as SendWarpMessageInput
+// assumes that [input] does not include selector (omits first 4 func signature bytes)
+func UnpackSendWarpMessageInput(input []byte) (SendWarpMessageInput, error) {
+	inputStruct := SendWarpMessageInput{}
+	err := WarpABI.UnpackInputIntoInterface(&inputStruct, "sendWarpMessage", input)
+
+	return inputStruct, err
+}
+
+// PackSendWarpMessage packs [inputStruct] of type SendWarpMessageInput into the appropriate arguments for sendWarpMessage.
+func PackSendWarpMessage(inputStruct SendWarpMessageInput) ([]byte, error) {
+	return WarpABI.Pack("sendWarpMessage", inputStruct.DestinationChainID, inputStruct.DestinationAddress, inputStruct.Payload)
+}
+
+// sendWarpMessage constructs an Avalanche Warp Message containing an AddressedPayload and emits a log to signal validators that they should
+// be willing to sign this message.
+func sendWarpMessage(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = contract.DeductGas(suppliedGas, SendWarpMessageGasCost); err != nil {
+		return nil, 0, err
+	}
+	// This gas cost includes buffer room because it is based off of the total size of the input instead of the produced payload.
+	// This ensures that we charge gas before we unpack the variable sized input.
+	payloadGas, overflow := math.SafeMul(SendWarpMessageGasCostPerByte, uint64(len(input)))
+	if overflow {
+		return nil, 0, vmerrs.ErrOutOfGas
+	}
+	if remainingGas, err = contract.DeductGas(remainingGas, payloadGas); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+	// unpack the arguments
```

Comment:
should `readOnly` be moved up the function so we don't have to deduct gas from remainingGas

Reply:
We could save gas in the case that someone calls this incorrectly in a `readOnly` state. I don't think this is a necessary change.